workflow/workflow.go

```diff
@@ -62,12 +64,59 @@ func (w *Workflow) process(trigger service.TriggerType, instanceHash hash.Hash,
 	for _, service := range services {
 		for _, wf := range service.Workflows {
 			if wf.Trigger.Match(trigger, instanceHash, key, data) {
-				_, err := w.execution.Execute(wf.Task.InstanceHash, eventHash, executionHash, wf.Task.TaskKey, data, []string{})
-				if err != nil {
+				if err := w.triggerExecution(wf, exec, eventHash, data); err != nil {
 					w.ErrC <- err
-					continue
 				}
 			}
 		}
 	}
 }
+
+func (w *Workflow) processExecution(exec *execution.Execution) error {
+	if exec.WorkflowHash.IsZero() {
+		return nil
+	}
+	wf, err := w.service.FindWorkflow(exec.WorkflowHash)
+	if err != nil {
+		return err
+	}
+	return w.triggerExecution(wf, exec, nil, exec.Outputs)
+}
+
+func (w *Workflow) triggerExecution(wf *service.Workflow, prev *execution.Execution, eventHash hash.Hash, data map[string]interface{}) error {
+	height, err := w.getHeight(prev)
+	if err != nil {
+		return err
+	}
+	if len(wf.Tasks) <= height {
+		// end of workflow
+		return nil
+	}
+	task := wf.Tasks[height]
+	hash, err := w.execution.Execute(wf.Hash, task.InstanceHash, eventHash, prev.Hash, task.TaskKey, data, []string{})
+	if err != nil {
+		return err
+	}
+	logrus.WithFields(logrus.Fields{
+		"workflow": wf.Key,
+		"task":     task.TaskKey,
+		"exec":     hash.String(),
+		"parent":   prev.Hash.String(),
+	}).Debug("workflow execution")
+	return nil
+}
+
+func (w *Workflow) getHeight(exec *execution.Execution) (int, error) {
+	if exec == nil {
+		return 0, nil
+	}
+	if exec.ParentHash == nil {
+		return 0, nil
+	}
+	parent, err := w.execution.Get(exec.ParentHash)
+	if err != nil {
+		return 0, err
+	}
+	parentHeight, err := w.getHeight(parent)
```

Comment:
possible stackoverflow if hash == parenthash. It shoudn't happen at all but we can add this check, what do you think?

Reply:
just put a panic in that case.